File: Python/_3_unit_tests/python_tests/suspension_elements/test_wishbone.py
from vehicle_model.suspension_model.suspension_elements._2_elements.wishbone import Wishbone
from vehicle_model.suspension_model.suspension_elements._1_elements.link import Link
from vehicle_model.suspension_model.suspension_elements._1_elements.node import Node
import numpy as np

from unittest import TestCase


class TestWishbone(TestCase):
    def test_wishbone_base_direction(self):
        inboard_fore = Node(position=[2, 0, 0])
        inboard_aft = Node(position=[0, 0, 0])
        outboard = Node(position=[1, 2, 0])

        fore_link = Link(inboard_node=inboard_fore, outboard_node=outboard)
        aft_link = Link(inboard_node=inboard_aft, outboard_node=outboard)

        wishbone = Wishbone(fore_link=fore_link, aft_link=aft_link)

        self.assertListEqual(wishbone.direction, [1, 0, 0])

    def test_wishbone_base_angle(self):
        inboard_fore = Node(position=[2, 0, 0])
        inboard_aft = Node(position=[0, 0, 0])
        outboard = Node(position=[1, 2, 0])

        fore_link = Link(inboard_node=inboard_fore, outboard_node=outboard)
        aft_link = Link(inboard_node=inboard_aft, outboard_node=outboard)

        wishbone = Wishbone(fore_link=fore_link, aft_link=aft_link)

        self.assertEqual(wishbone.angle, 0)
    
    def test_rotate(self):
        inboard_fore = Node(position=[2, 0, 0])
        inboard_aft = Node(position=[0, 0, 0])
        outboard = Node(position=[1, 2, 0])

        fore_link = Link(inboard_node=inboard_fore, outboard_node=outboard)
        aft_link = Link(inboard_node=inboard_aft, outboard_node=outboard)

        wishbone = Wishbone(fore_link=fore_link, aft_link=aft_link)
        wishbone.rotate(angle=np.pi/2)

        self.assertListEqual([round(x, 7) for x in wishbone.fore_link.outboard_node.position], [1, 0, 2])
    
    def test_child_rotate(self):
        inboard_fore = Node(position=[2, 0, 0])
        inboard_aft = Node(position=[0, 0, 0])
        outboard = Node(position=[1, 2, 0])

        fore_link = Link(inboard_node=inboard_fore, outboard_node=outboard)
        aft_link = Link(inboard_node=inboard_aft, outboard_node=outboard)

        wishbone = Wishbone(fore_link=fore_link, aft_link=aft_link)

        child_node = Node(position=[1, 1, 0])
        outboard.add_child(node=child_node)
        
        wishbone.rotate(angle=np.pi/2)

        self.assertListEqual([round(x, 7) for x in child_node.position], [1, 0, 1])

    def test_plane_xy(self):
        inboard_fore = Node(position=[2, 0, 0])
        inboard_aft = Node(position=[0, 0, 0])
        outboard = Node(position=[1, 2, 0])

        fore_link = Link(inboard_node=inboard_fore, outboard_node=outboard)
        aft_link = Link(inboard_node=inboard_aft, outboard_node=outboard)

        wishbone = Wishbone(fore_link=fore_link, aft_link=aft_link)
        a, b, c, x_0, y_0, z_0 = wishbone.plane

        # Check if planes are equivalent type shit
        point_1 = a * (1 - x_0) + b * (0 - y_0) + c * (0 - z_0)
        point_2 = a * (0 - x_0) + b * (1 - y_0) + c * (0 - z_0)
        point_3 = a * (0 - x_0) + b * (0 - y_0) + c * (0 - z_0)

        self.assertListEqual([point_1, point_2, point_3], [0, 0, 0])

    def test_direction_vec(self):
        inboard_fore = Node(position=[2, 0, 0])
        inboard_aft = Node(position=[0, 0, 0])
        outboard = Node(position=[1, 2, 0])

        fore_link = Link(inboard_node=inboard_fore, outboard_node=outboard)
        aft_link = Link(inboard_node=inboard_aft, outboard_node=outboard)

        wishbone = Wishbone(fore_link=fore_link, aft_link=aft_link)

        self.assertListEqual(wishbone.direction_vec, [1, 0, 0])

File: unit_tests/cpp_tests/test_wishbone.cpp
#include "test_wishbone.h"
#include <cmath>

using namespace blaze;

// Test: Wishbone base direction
TEST_F(TestWishbone, test_wishbone_base_direction) {
    Node inboard_fore(StaticVector<double, 3UL>{2, 0, 0});
    Node inboard_aft(StaticVector<double, 3UL>{0, 0, 0});
    Node outboard(StaticVector<double, 3UL>{1, 2, 0});
    
    Beam fore_link(&inboard_fore, &outboard);
    Beam aft_link(&inboard_aft, &outboard);
    
    Wishbone wishbone(&fore_link, &aft_link);
    
    StaticVector<double, 3UL> expected = {1, 0, 0};
    expectVectorEqual(wishbone.getDirection(), expected);
}

// Test: Wishbone base angle
TEST_F(TestWishbone, test_wishbone_base_angle) {
    Node inboard_fore(StaticVector<double, 3UL>{2, 0, 0});
    Node inboard_aft(StaticVector<double, 3UL>{0, 0, 0});
    Node outboard(StaticVector<double, 3UL>{1, 2, 0});
    
    Beam fore_link(&inboard_fore, &outboard);
    Beam aft_link(&inboard_aft, &outboard);
    
    Wishbone wishbone(&fore_link, &aft_link);
    
    EXPECT_DOUBLE_EQ(wishbone.getAngle(), 0.0);
}

// Test: Rotate wishbone
TEST_F(TestWishbone, test_rotate) {
    Node inboard_fore(StaticVector<double, 3UL>{2, 0, 0});
    Node inboard_aft(StaticVector<double, 3UL>{0, 0, 0});
    Node outboard(StaticVector<double, 3UL>{1, 2, 0});
    
    Beam fore_link(&inboard_fore, &outboard);
    Beam aft_link(&inboard_aft, &outboard);
    
    Wishbone wishbone(&fore_link, &aft_link);
    wishbone.rotate(M_PI / 2.0);
    
    StaticVector<double, 3UL> expected = {1, 0, 2};
    StaticVector<double, 3UL> actual = fore_link.getOutboardNode()->position;
    expectVectorNear(actual, expected, 1e-6);
}

// Test: Child node rotation
TEST_F(TestWishbone, test_child_rotate) {
    Node inboard_fore(StaticVector<double, 3UL>{2, 0, 0});
    Node inboard_aft(StaticVector<double, 3UL>{0, 0, 0});
    Node outboard(StaticVector<double, 3UL>{1, 2, 0});
    
    Beam fore_link(&inboard_fore, &outboard);
    Beam aft_link(&inboard_aft, &outboard);
    
    Wishbone wishbone(&fore_link, &aft_link);
    
    Node child_node(StaticVector<double, 3UL>{1, 1, 0});
    outboard.add_child(&child_node);
    
    wishbone.rotate(M_PI / 2.0);
    
    StaticVector<double, 3UL> expected = {1, 0, 1};
    StaticVector<double, 3UL> actual = child_node.position;
    expectVectorNear(actual, expected, 1e-6);
}

// Test: Plane calculation (XY plane)
TEST_F(TestWishbone, test_plane_xy) {
    Node inboard_fore(StaticVector<double, 3UL>{2, 0, 0});
    Node inboard_aft(StaticVector<double, 3UL>{0, 0, 0});
    Node outboard(StaticVector<double, 3UL>{1, 2, 0});
    
    Beam fore_link(&inboard_fore, &outboard);
    Beam aft_link(&inboard_aft, &outboard);
    
    Wishbone wishbone(&fore_link, &aft_link);
    StaticVector<double, 6UL> plane_params = wishbone.plane();
    
    double a = plane_params[0];
    double b = plane_params[1];
    double c = plane_params[2];
    double x_0 = plane_params[3];
    double y_0 = plane_params[4];
    double z_0 = plane_params[5];
    
    // Check if planes are equivalent - test points should satisfy plane equation
    double point_1 = a * (1 - x_0) + b * (0 - y_0) + c * (0 - z_0);
    double point_2 = a * (0 - x_0) + b * (1 - y_0) + c * (0 - z_0);
    double point_3 = a * (0 - x_0) + b * (0 - y_0) + c * (0 - z_0);
    
    EXPECT_NEAR(point_1, 0.0, 1e-6);
    EXPECT_NEAR(point_2, 0.0, 1e-6);
    EXPECT_NEAR(point_3, 0.0, 1e-6);
}

// Test: Direction vector
TEST_F(TestWishbone, test_direction_vec) {
    Node inboard_fore(StaticVector<double, 3UL>{2, 0, 0});
    Node inboard_aft(StaticVector<double, 3UL>{0, 0, 0});
    Node outboard(StaticVector<double, 3UL>{1, 2, 0});
    
    Beam fore_link(&inboard_fore, &outboard);
    Beam aft_link(&inboard_aft, &outboard);
    
    Wishbone wishbone(&fore_link, &aft_link);
    
    StaticVector<double, 3UL> expected = {1, 0, 0};
    StaticVector<double, 3UL> actual = wishbone.direction_vec();
    expectVectorEqual(actual, expected);
}

